pkg/metrics/builder/builder_test.go

```diff
@@ -0,0 +1,226 @@
+		},
+		{
+			name:          "Test invalid labels",
+			collectorType: "histogram",
+			opts: MetricOpts{
+				labels:         nil,
+				labelKeyValues: map[string]string{"test1": "test1", "test2": "test2"},
+				desc:           "test metric",
+				name:           "test",
+				metricType:     "Histogram",
+			},
+			expectedBuilderError:       true,
+			expectedPromCollectorError: false,
+			expectedErrorMessage:       `labels cannot be empty`,
+		},
+		{
+			name:          "Test empty labels",
+			collectorType: "histogram",
+			opts: MetricOpts{
+				labels:         []string{},
+				labelKeyValues: map[string]string{"test1": "test1", "test2": "test2"},
+				desc:           "test metric",
+				name:           "test",
+				metricType:     "Histogram",
+			},
+			expectedBuilderError:       true,
+			expectedPromCollectorError: false,
+			expectedErrorMessage:       `labels cannot be empty`,
+		},
+		{
+			name:          "Test invalid name",
+			collectorType: "histogram",
+			opts: MetricOpts{
+				labels:         []string{"test1", "test2"},
+				labelKeyValues: map[string]string{"test1": "test1", "test2": "test2"},
+				desc:           "test metric",
+				name:           "",
+				metricType:     "Histogram",
+			},
+			expectedBuilderError:       true,
+			expectedPromCollectorError: false,
+			expectedErrorMessage:       `name cannot be empty`,
+		},
+		{
+			name:          "Test invalid metric type",
+			collectorType: "histogram",
+			opts: MetricOpts{
+				labels:         []string{"test1", "test2"},
+				labelKeyValues: map[string]string{"test1": "test1", "test2": "test2"},
+				desc:           "test metric",
+				name:           "test",
+				metricType:     "",
+			},
+			expectedBuilderError:       true,
+			expectedPromCollectorError: false,
+			expectedErrorMessage:       `metricType cannot be empty`,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			metricBuilder, err := NewMetricBuilder(tc.opts)
+			if err != nil {
+				if tc.expectedBuilderError {
+					assert.EqualValues(t, tc.expectedErrorMessage, err.Error())
+					return
+				}
+				assert.Failf(t, err.Error(), "expected successful builder creation")
+				return
+			} else if tc.expectedBuilderError {
+				assert.Fail(t, "expected builder creation failure but builder was created")
+				return
+			}
+
+			collector, err := metricBuilder.PromCollector()
+			if err != nil {
+				if tc.expectedPromCollectorError {
+					assert.EqualValues(t, tc.expectedErrorMessage, err.Error())
+					return
+				}
+				assert.Failf(t, err.Error(), "expected successful builder creation")
+				return
+			} else if tc.expectedPromCollectorError {
+				assert.Fail(t, "expected builder creation failure but builder was created")
+				return
+			}
+
+			assert.EqualValues(t, tc.collectorType, reflect.TypeOf(collector).Elem().Name())
+			expectedString := getDescStruct(tc.opts)
+			assert.EqualValues(t, expectedString, getCollectorDescription(collector))
```

Comment:
Why are you testing how prom lib creates description? That seems like not providing us any useful unit testing.

Reply:
The motivation behind testing that was I wanted to see if the Collector object actually has the values that were in the MetricBuilder and all the Collector's properties are private except a String. It's a roundabout way to test if the values are being set but I think it's the only way as all are private. The other way was to use channels and pick the values when they are being set but I thought that would be overkill.